Remove all bashisms (resolves #53)

aimileus opened the pull request:
I tried to remove all bashisms, and now checkbashisms does not complain anymore:

- POSIX only supports single digit file descriptors. I used 3 instead of 42 as the file descriptor for /dev/null.
- echo -e replaces certain backslash expressions. Since, there were none of them, I just removed the -e flag. The `${variable}` syntax should still work.
- POSIX does not set the $UID environment variable. I replaced it with the custom variable $uid.
- The $'' syntax for backslash expressions does not exists in POSIX.  Since it was only used for the tab character, I just replaced it with the custom $tab variable at the beginning of the script.

In my (limited) testing, create, enter and list still works, using both bash and dash as interpreter. The --verbose option also still works, although I might have forgotten to change the file descriptor somewhere.

debarshiray commented:
Thanks for the excellent patch! Sadly, I ran into a ton of conflicts when trying to rebase it. In the end it seemed like it would be easier to cherry-pick each of the changes manually on top of master.

I have now created pull request #57 for the *single digit file descriptors* change. I hope I got it right.

I can slowly work through the rest of the changes in your patch, but feel free to rebase it yourself, if you prefer.

Once again, thanks for your contribution!

debarshiray commented:
By the way, you've got a lot of changes here. Feel free to add yourself to the copyright notice. :)

aimileus commented:
I rebased on master and it seems to work now. And although I changed a lot of lines of code, I do not feel I deserve a place in the copyright notice by replacing &42 by &3. Anyway, thank you very much for the offer!

debarshiray commented:
Sorry, it took me a fortnight to get back to this. I have now merged the `$UID` replacement via #72

aimileus commented:
Thanks! I rebased on master again, so you can apply it cleanly, if you want to.

debarshiray commented on toolbox:
```diff
@@ -529,15 +530,15 @@ list_containers()
 
     if [ "$output" != "" ]; then
-        echo -e "${LBC}Containers created by toolbox${NC}"
+        echo "${LBC}Containers created by toolbox${NC}"
```
I was trying to read up on why the `-e` doesn't matter. The `man` pages for both the bash built-in in `bash(1)` and the external binary in `echo(1)` say that when `-e` is in effect the `\0NNN` sequence is one of the recognized sequences. Wouldn't that cover this case?

Or am I reading it wrong? Another option is to use `printf`.

aimileus replied:
First of all, sorry for the delay. Yes, you are correct, I just removed the -e flag and it seemed to work fine, but apparently I was incorrect. I pushed a new commit that uses printf.

debarshiray replied:
No need to apologize. :) Thanks for updating the patch.

debarshiray commented:
I merged the `$'...'` replacement via #77

debarshiray commented on toolbox:
```diff
@@ -775,7 +775,7 @@ list_containers()
                       is_running=$($prefix_sudo podman inspect "$id" --format "{{.State.Running}}" 2>&3)
                       if $is_running; then
-                          echo -e "${LGC}$container${NC}"
+                          echo "${LGC}$container${NC}"
```
We need to replace this `echo -e` with `printf` too.

(By the way, when I tested using `echo` without the `-e`, Bash refused to intepret the terminal escape sequences as control characters and printed them as is. So, maybe the behaviour that you saw was specific to Dash.)

debarshiray replied:
I took the liberty to make the above adjustments and merged to `master`. Thanks for your contribution!

aimileus replied:
Thank you for merging!
